### PR TITLE
Early program termination only from main()

### PR DESCRIPTION
**Pull request by vasvolodimir**
Fix for https://github.com/htop-dev/htop/issues/331

### CODE REVIEW

File: darwin/Platform.c [hunk 1]
```diff
@@ -126,7 +126,7 @@ static double Platform_nanosecondsPerMachTick = 1.0;
 
 static double Platform_nanosecondsPerSchedulerTick = -1;
 
-void Platform_init(void) {
+int Platform_init(void) {
```
Comment (BenBE) [inline darwin/Platform.c:129]: Why change this, if the only return value is `0` anyways? If this shares the same semantics as `parseArguments` then a similar (or the same) enum should be used for marking the return value.

File: darwin/Platform.c [hunk 2]
```diff
@@ -139,6 +139,8 @@ void Platform_init(void) {
 
    const double nanos_per_sec = 1e9;
    Platform_nanosecondsPerSchedulerTick = nanos_per_sec / scheduler_ticks_per_sec;
+
+   return 0;
```
Comment (BenBE) [inline darwin/Platform.c:143]: This return value doesn't make much sense, in particular as an int. IMHO `bool` (`stdbool.h`) would be better suited.

File: CommandLine.c [hunk 3]
```diff
@@ -283,12 +283,17 @@ int CommandLine_run(const char* name, int argc, char** argv) {
    else
       setlocale(LC_CTYPE, "");
 
-   CommandLineSettings flags = parseArguments(name, argc, argv);
+   int status;
+   CommandLineSettings flags;
+
+   if ((status = parseArguments(name, argc, argv, &flags)))
+      return status < 0;
```
Comment (BenBE) [inline CommandLine.c:290]: This does an implicit `bool`-to-`int` conversion and is kinda crude IMO.

File: CommandLine.c [hunk 4]
```diff
@@ -130,10 +130,10 @@ static CommandLineSettings parseArguments(const char* program, int argc, char**
       switch (opt) {
          case 'h':
             printHelpFlag(program);
-            exit(0);
+            return 1;
```
Comment (BenBE) [inline CommandLine.c:133]: Avoid magic numbers. Define a small enum for proper initialization outcomes and return the appropriate constant instead.

File: pcp/Platform.c [hunk 5]
```diff
@@ -309,7 +309,8 @@ void Platform_init(void) {
    sts = pmLookupName(pcp->totalMetrics, pcp->names, pcp->pmids);
    if (sts < 0) {
       fprintf(stderr, "Error: cannot lookup metric names: %s\n", pmErrStr(sts));
-      exit(1);
+      Platform_done();
```
Comment (BenBE) [inline pcp/Platform.c:312]: @natoscott Can you take a look here for PCP? Not sure if this call to `Platform_done()` is correct here.

**Review by BenBE — CHANGES_REQUESTED** [review]

**Comment by vasvolodimir** [comment]
Hi @BenBE,
Rebase has done.

File: CommandLine.c [hunk 6]
```diff
@@ -288,12 +290,17 @@ int CommandLine_run(const char* name, int argc, char** argv) {
    else
       setlocale(LC_CTYPE, "");
 
-   CommandLineSettings flags = parseArguments(name, argc, argv);
+   CommandLineStatus status;
+   CommandLineSettings flags;
```
Comment (BenBE) [inline CommandLine.c:294]: Better to not leave things uninitialized here.
```suggestion
   CommandLineStatus status = STATUS_OK;
   CommandLineSettings flags = 0;
```

File: CommandLine.c [hunk 7]
```diff
@@ -288,12 +290,17 @@ int CommandLine_run(const char* name, int argc, char** argv) {
...
+
+   if ((status = parseArguments(name, argc, argv, &flags)) != STATUS_OK)
+      return status != STATUS_OK_EXIT ? 1 : 0;
 
    if (flags.readonly)
       Settings_enableReadonly();
 
-   Platform_init();
+   if (Platform_init() == false)
```
Comment (BenBE) [inline CommandLine.c:302]:
```suggestion
   if (!Platform_init())
```

**Review by BenBE — CHANGES_REQUESTED** [review]

**Review by BenBE — APPROVED** [review]

**Pull request merged**